### PR TITLE
add pypy3.7 CI run

### PR DESCRIPTION
**Pull request by mattip**
Putting this out there as a test to see if this works.

### CODE REVIEW

**Comment by mattip** [comment]
Nope, it [failed on my fork](https://github.com/mattip/psycopg3/runs/2879277582?check_suite_focus=true#step:6:76) since typed-ast has declared [they can not support PyPy](https://github.com/python/typed_ast/issues/111#issuecomment-493279393). Is there a way to work around that, by disabling mypy checks for this CI run?

**Comment by mattip** [comment]
I will ping here when I get this working on my fork

File: tests/pool/test_pool.py [hunk 1]
```diff
@@ -12,6 +13,11 @@
 from psycopg3.pq import TransactionStatus
 
 
+def gc_collect():
```
Comment (dvarrazzo) [inline tests/pool/test_pool.py:16]: This seems useful, as the gc tests seem somewhat flaky on cpython too.

Comment (dvarrazzo) [reply tests/pool/test_pool.py:16]: I have incorporated a similar changeset in master.

**Comment by thedrow** [comment]
@mattip Could you rebase this PR?

File: psycopg3/setup.py [hunk 2]
```diff
@@ -32,7 +32,7 @@
         f"psycopg3-binary == {version}",
     ],
     "test": [
-        "mypy >= 0.812",
+        "mypy >= 0.812 ; implementation_name!='pypy'",
```
Comment (dvarrazzo) [inline psycopg3/setup.py:35]: Please use consistent spacing around the operator

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -17,6 +17,8 @@ jobs:
             postgres: 10
           - python: 3.7
             postgres: 11
+          - python: pypy-3.7
+            postgres: 11
```
[the remark this answers is not in the document]
Comment (mattip) [reply .github/workflows/tests.yml]: changed to 13

File: psycopg3/setup.py (hunk 2 above)
Comment (mattip) [reply psycopg3/setup.py]: removed space before ';'

**Review by dvarrazzo — CHANGES_REQUESTED** [review]

File: tests/test_copy.py [hunk 4]
```diff
@@ -583,9 +589,14 @@ def work():
                 gc_collect()
                 n.append(len(gc.get_objects()))
 
-            assert (
-                n[0] == n[1] == n[2]
-            ), f"objects leaked: {n[1] - n[0]}, {n[2] - n[1]}"
+            if sys.implementation.name == "pypy":
+                assert (
+                    n[0] >= n[1] >= n[2]
+                ), f"objects leaked: {n[1] - n[0]}, {n[2] - n[1]}"
```
[the remark this answers is not in the document]
Comment (mattip) [reply tests/test_copy.py:595]: Sorry, I think this came from a bad merge. fixing.

File: psycopg3/setup.py (hunk 2 above)
Comment (dvarrazzo) [reply psycopg3/setup.py]: I meant the spaces around the operator, as you used in the other line :)

**Comment by mattip** [comment]
Workflow is awaiting approval to run since I am a new contributor

File: tests/test_cursor.py [hunk 5]
```diff
@@ -13,6 +14,11 @@
 from .utils import gc_collect
 
 
+def gc_collect():
```
Comment (dvarrazzo) [inline tests/test_cursor.py:17]: Note that now there is a definition of `gc_collect()` in `.utils`, see line above

**Comment by dvarrazzo** [comment]
@mattip 

> Workflow is awaiting approval to run since I am a new contributor

I have poked a bit at the settings, you should be able to run workflows without approval needed now

**Comment by mattip** [comment]
Three failures:
```
FAILED tests/test_connection.py::test_connection_warn_close - assert not Warn...
FAILED tests/test_connection_async.py::test_connection_warn_close - Assertion...
FAILED tests/test_server_cursor_async.py::test_warn_close - psycopg.errors.In...
```

The first warns when it shouldn't, the second does not warn when it should (`gc_collect` might help) and I don't know what is going on in the third one.

**Comment by dvarrazzo** [comment]
I think those tests about resources warnings should be just skipped on pypi: they assume the behaviour of the cpython garbage collector.

In order to see a clearer error message for the third, please rebase on master to include https://github.com/psycopg/psycopg/commit/c8e895281151a95da607541515e07e164f5fac7b

File: tests/pool/test_pool.py [hunk 6]
```diff
@@ -12,6 +13,11 @@
 from psycopg.pq import TransactionStatus
 
 
+def gc_collect():
+    for i in range(3):
+        gc.collect()
```
Comment (dvarrazzo) [inline tests/pool/test_pool.py:18]: This can be imported from the `.utils` module too

File: psycopg/setup.py [hunk 7]
```diff
@@ -37,7 +37,7 @@
     ],
     # Requirements to run the test suite
     "test": [
-        "mypy >= 0.812",
+        "mypy >= 0.812; implementation_name !='pypy'",
```
Comment (dvarrazzo) [inline psycopg/setup.py:40]: Still not formatted ok :)

**Review by dvarrazzo — CHANGES_REQUESTED** [review]

File: psycopg/setup.py (hunk 7 above)
Comment (mattip) [reply psycopg/setup.py:40]: do I need a space before the semicolon?

Comment (dvarrazzo) [reply psycopg/setup.py:40]:
```suggestion
        "mypy >= 0.812; implementation_name != 'pypy'",
```

File: tests/test_connection.py [hunk 8]
```diff
@@ -89,19 +89,25 @@ def test_broken(conn):
     assert conn.broken
 
 
+@pytest.mark.skipif(
```
Comment (dvarrazzo) [inline tests/test_connection.py:92]: If you are skipping this test on pypy, please drop the `gc_collect()` you added

**Review by dvarrazzo — CHANGES_REQUESTED** [review]

File: tests/pool/test_pool.py [hunk 9]
```diff
@@ -11,6 +11,8 @@
 import psycopg_pool as pool
 from psycopg.pq import TransactionStatus
 
+from .utils import gc_collect
```
Comment (dvarrazzo) [inline tests/pool/test_pool.py:14]:
```suggestion
from ..utils import gc_collect
```

**Comment by mattip** [comment]
I don't understand the `psycopg.errors.InFailedSqlTransaction: current transaction is aborted, commands ignored until end of transaction block` error. Is that supposed to raise differently or not happen at all?

**Comment by dvarrazzo** [comment]
> I don't understand the psycopg.errors.InFailedSqlTransaction: current transaction is aborted, commands ignored until end of transaction block error. Is that supposed to raise differently or not happen at all?

Ah, it's still in `tests/test_server_cursor.py:99`? That' strange. I assumed the test could have been put in bad state by an assert but probably it's not the case.

You can try and drop the retry loop to see what error happen. If you find and fix it, probably the transaction_status check I added in c8e8952 is not needed and can be removed.

**Comment by om26er** [comment]
I would like to help with this if there is a chance we can revive this effort. We currently use `asyncpg` with SQLAlchemy but PyPy support is a hard requirement for us and there is little hope in `asyncpg` getting that. OTH, psycopg seems to almost work https://github.com/psycopg/psycopg/issues/391#issuecomment-1368471790 -- My first goal is to get psycopg working with ctypes on PyPy and then as a next step, I would like to explore cffi support.

**Comment by mattip** [comment]
Feel free to start over from the beginning with addding pypy to the CI. Note that the PyPy team is no longer developing pypy3.7, and pypy3.8 is also nearing EOL, so it would be best to do this with pypy3.9.

**Comment by ngnpope** [comment]
Superseded by #686.

**Pull request closed without merging**